x-pack/plugins/security_solution/server/endpoint/lib/artifacts/manifest.ts

```diff
@@ -62,15 +70,70 @@ export class Manifest {
     return manifest;
   }
 
+  public static fromPkgConfig(manifestPkgConfig: ManifestSchema): Manifest | null {
+    if (manifestSchema.is(manifestPkgConfig)) {
+      const manifest = new Manifest(manifestPkgConfig.schema_version);
+      for (const [identifier, artifactRecord] of Object.entries(manifestPkgConfig.artifacts)) {
+        const artifact = {
+          identifier,
+          compressionAlgorithm: artifactRecord.compression_algorithm,
+          encryptionAlgorithm: artifactRecord.encryption_algorithm,
+          decodedSha256: artifactRecord.decoded_sha256,
+          decodedSize: artifactRecord.decoded_size,
+          encodedSha256: artifactRecord.encoded_sha256,
+          encodedSize: artifactRecord.encoded_size,
+        };
+        manifest.addEntry(artifact);
+      }
+      return manifest;
+    } else {
+      return null;
+    }
+  }
+
+  public async compressArtifact(id: string): Promise<Error | null> {
+    try {
+      const artifact = this.getArtifact(id);
+      if (artifact == null) {
+        throw new Error(`Corrupted manifest detected. Artifact ${id} not in manifest.`);
+      }
+
+      const compressedArtifact = await maybeCompressArtifact(artifact);
+      if (!isCompressed(compressedArtifact)) {
+        throw new Error(`Unable to compress artifact: ${id}`);
+      } else if (!internalArtifactCompleteSchema.is(compressedArtifact)) {
+        throw new Error(`Incomplete artifact detected: ${id}`);
+      }
+      this.addEntry(compressedArtifact);
+    } catch (err) {
+      return err;
+    }
+    return null;
+  }
+
+  public equals(manifest: Manifest): boolean {
+    return this.getSha256() === manifest.getSha256();
+  }
+
+  public getSha256(): string {
+    let sha256 = createHash('sha256');
+    Object.keys(this.entries)
+      .sort()
```

Comment:
If we are sorting before generating the hash, won't we need to make sure that the sort is persisted since the Endpoint will have to hash as well to verify the package?

Reply:
@alexk307 The endpoint doesn't verify this hash. This is just for us to know whether or not a policy needs to be updated (if the manifest hashes don't match, we need to update the policy).